File: skypydb/utils/distance_metrics.py
"""Distance helpers used during vector similarity search."""

import math
from typing import List


def cosine_similarity(vec1: List[float], vec2: List[float]) -> float:
    """Return cosine similarity between two vectors."""

    if len(vec1) != len(vec2):
        raise ValueError(f"Vector dimensions don't match: {len(vec1)} vs {len(vec2)}")

    dot_product = sum(a * b for a, b in zip(vec1, vec2))
    norm1 = math.sqrt(sum(a * a for a in vec1))
    norm2 = math.sqrt(sum(b * b for b in vec2))
    if norm1 == 0 or norm2 == 0:
        return 0.0
    return dot_product / (norm1 * norm2)


def euclidean_distance(vec1: List[float], vec2: List[float]) -> float:
    """Return Euclidean distance between two vectors."""

    if len(vec1) != len(vec2):
        raise ValueError(f"Vector dimensions don't match: {len(vec1)} vs {len(vec2)}")
    return math.sqrt(sum((a - b) ** 2 for a, b in zip(vec1, vec2)))

File: skypydb-js/src/utils/distance_metrics.ts
export function cosine_similarity(vec1: number[], vec2: number[]): number {
  if (vec1.length !== vec2.length) {
    throw new Error(
      `Vector dimensions don't match: ${vec1.length} vs ${vec2.length}`,
    );
  }

  let dot = 0;
  let norm1 = 0;
  let norm2 = 0;
  for (let index = 0; index < vec1.length; index += 1) {
    const a = vec1[index];
    const b = vec2[index];
    dot += a * b;
    norm1 += a * a;
    norm2 += b * b;
  }
  if (norm1 === 0 || norm2 === 0) {
    return 0;
  }
  return dot / (Math.sqrt(norm1) * Math.sqrt(norm2));
}

export function euclidean_distance(vec1: number[], vec2: number[]): number {
  if (vec1.length !== vec2.length) {
    throw new Error(
      `Vector dimensions don't match: ${vec1.length} vs ${vec2.length}`,
    );
  }
  let sum = 0;
  for (let index = 0; index < vec1.length; index += 1) {
    const delta = vec1[index] - vec2[index];
    sum += delta * delta;
  }
  return Math.sqrt(sum);
}
